src/main/java/org/terasology/logic/inventory/InventoryUIClientSystem.java

```diff
@@ -69,4 +91,121 @@ public void onToggleInventory(InventoryButton event, EntityRef entity, ClientCom
             // do not consume the event, so that the inventory will still open
         }
     }
+
+
+    /*
+      The numbersBetween() and getTransferEntity() methods were
+      originally in the InventoryCell class. They were copied over
+      to here because they are private functions that the
+      moveItemSmartly() method needs to function. The first section
+      of code in onClosed() is based on the moveItemSmartly() method.
+    */
+    private List<Integer> numbersBetween(int start, int exclusiveEnd) {
+        List<Integer> numbers = new ArrayList<>();
+        for (int number = start; number < exclusiveEnd; number++) {
+            numbers.add(number);
+        }
+        return numbers;
+    }
+
+    private EntityRef getTransferEntity() {
+        return localPlayer.getCharacterEntity().getComponent(CharacterComponent.class).movingItem;
+    }
+
+    @Override
+    public void preAutoSave(){
+        /*
+          The code below was originally taken from moveItemSmartly() in
+          InventoryCell.class and slightly modified to work here.
+
+          The way items are being moved to and from the hotbar is really
+          similar to what was needed here to take them out of the transfer
+          slot and sort them into the inventory.
+        */
+        EntityRef playerEntity = localPlayer.getCharacterEntity();
+        EntityRef movingItemSlot = playerEntity.getComponent(CharacterComponent.class).movingItem;
+
+        movingItemItem = invManager.getItemInSlot(movingItemSlot, 0);
+        
+        movingItemCount = invManager.getStackSize(movingItemItem);
+
+        EntityRef fromEntity = movingItemSlot;
+        int fromSlot = 0;
+
+        InventoryComponent playerInventory = playerEntity.getComponent(InventoryComponent.class);
+
+        if (movingItemItem != EntityRef.NULL) {
+
+
+
+            if (playerInventory == null) {
+                return;
+            }
+            CharacterComponent characterComponent = playerEntity.getComponent(CharacterComponent.class);
+            if (characterComponent == null) {
+                logger.error("Character entity of player had no character component");
+                return;
+            }
+            int totalSlotCount = playerInventory.itemSlots.size();
+
+            EntityRef interactionTarget = characterComponent.predictedInteractionTarget;
+            InventoryComponent interactionTargetInventory = interactionTarget.getComponent(InventoryComponent.class);
+
+
+            EntityRef targetEntity;
+            List<Integer> toSlots = new ArrayList<>(totalSlotCount);
+            if (fromEntity.equals(playerEntity)) {
+
+                if (interactionTarget.exists() && interactionTargetInventory != null) {
+                    targetEntity = interactionTarget;
+                    toSlots = numbersBetween(0, interactionTargetInventory.itemSlots.size());
+                } else {
```

Comment:
This `else` branch is exactly the same as the next one, I think they should be combined.